core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedDocServicePlugin.java

```diff
@@ -399,33 +452,31 @@ static ServiceSpecification generate(Map<Class<?>, String> serviceDescription,
                 })
                 .collect(toImmutableSet());
 
-        return ServiceSpecification.generate(serviceInfos, AnnotatedDocServicePlugin::newNamedTypeInfo);
+        return ServiceSpecification.generate(
+                serviceInfos, typeSignature -> newNamedTypeInfo(typeSignature, namedTypeInfoProvider));
     }
 
-    private static NamedTypeInfo newNamedTypeInfo(TypeSignature typeSignature) {
-        final Class<?> type = (Class<?>) typeSignature.namedTypeDescriptor();
-        if (type == null) {
+    private static NamedTypeInfo newNamedTypeInfo(TypeSignature typeSignature, NamedTypeInfoProvider provider) {
+        final Object typeDescriptor = typeSignature.namedTypeDescriptor();
+        if (typeDescriptor == null) {
             throw new IllegalArgumentException("cannot create a named type from: " + typeSignature);
         }
 
-        if (type.isEnum()) {
-            @SuppressWarnings("unchecked")
-            final Class<? extends Enum<?>> enumType = (Class<? extends Enum<?>>) type;
-            return new EnumInfo(enumType);
+        NamedTypeInfo namedTypeInfo = provider.newNamedTypeInfo(typeDescriptor);
+        if (namedTypeInfo != null) {
+            return namedTypeInfo;
         }
 
-        return newStructInfo(type);
-    }
-
-    private static StructInfo newStructInfo(Class<?> structClass) {
-        final String name = structClass.getName();
-
-        final Field[] declaredFields = structClass.getDeclaredFields();
-        final List<FieldInfo> fields =
-                Stream.of(declaredFields)
-                      .map(f -> FieldInfo.of(f.getName(), toTypeSignature(f.getGenericType())))
-                      .collect(Collectors.toList());
-        return new StructInfo(name, fields);
+        // If the type can be serialized to JSON, try extracting a `StructInfo` using Jackson.
+        // Don't care about the request object because the `StructInfo` for the request object is processed
+        // when building the `FieldInfo` of the `MethodInfo`.
+        namedTypeInfo = responseJsonNamedTypeInfoProvider.newNamedTypeInfo(typeDescriptor);
```

Comment:
Question) Technically, couldn't types also be created as a named TypeSignature (container types, enum types, etc..)? Would it be correct behavior to use `requestJsonNamedTypeInfoProvider` instead in this case?

e.g. https://github.com/line/armeria/blob/9298e47a3262708ee5bed935e97079e2a8ca4ebf/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedDocServicePlugin.java#L295

If a method parameter were for instance `List<JsonRequest> ints`

Reply:
Taking a second look, maybe this is more of a problem of the UI. Our docs client assumes that structs are the same format whether they are for requests or responses, but this PR starts to make a distinction.